pkg/nimble/netif/nimble_netif.c

```diff
@@ -0,0 +1,621 @@
+#include "nimble_netif_conn.h"
+#include "nimble_riot.h"
+#include "host/ble_gap.h"
+#include "host/util/util.h"
+
+#define ENABLE_DEBUG            (0)
+#include "debug.h"
+
+#ifdef MODULE_GNRC_SIXLOWPAN
+#define NETTYPE                 GNRC_NETTYPE_SIXLOWPAN
+#else
+#define NETTYPE                 GNRC_NETTYPE_UNDEF
+#endif
+
+/* maximum packet size for IPv6 packets */
+#ifndef NIMBLE_NETIF_IPV6_MTU
+#define NIMBLE_NETIF_IPV6_MTU    (1280U)    /* as specified in RFC7668 */
+#endif
+
+/* buffer configuration
+ * - we need one RX and one TX buffer per connection */
+#define MTU_SIZE                (NIMBLE_NETIF_IPV6_MTU)
+#define MBUF_OVHD               (sizeof(struct os_mbuf) + \
+                                 sizeof(struct os_mbuf_pkthdr))
+#define MBUF_SIZE               (MBUF_OVHD + MYNEWT_VAL_BLE_L2CAP_COC_MPS)
+#define MBUF_CNT                (MYNEWT_VAL_BLE_MAX_CONNECTIONS * 2 * \
+                                 ((MTU_SIZE + (MBUF_SIZE - 1)) / MBUF_SIZE))
+
+/* thread flag used for signaling transmit readiness */
+#define FLAG_TX_UNSTALLED       (1u << 13)
+
+/* allocate a stack for the netif device */
+static char _stack[THREAD_STACKSIZE_DEFAULT];
+static thread_t *_netif_thread;
+
+/* keep the actual device state */
+static gnrc_netif_t *_nimble_netif = NULL;
+static gnrc_nettype_t _nettype = NETTYPE;
+
+/* keep a reference to the event callback */
+static nimble_netif_eventcb_t _eventcb;
+
+/* allocation of memory for buffering IP packets when handing them to NimBLE */
+static os_membuf_t _mem[OS_MEMPOOL_SIZE(MBUF_CNT, MBUF_SIZE)];
+static struct os_mempool _mem_pool;
+static struct os_mbuf_pool _mbuf_pool;
+
+/* notify the user about state changes for a connection context */
+static void _notify(int handle, nimble_netif_event_t event)
+{
+    if (_eventcb) {
+        _eventcb(handle, event);
+    }
+}
+
+static void _netif_init(gnrc_netif_t *netif)
+{
+    (void)netif;
+
+    /* save the threads context pointer, so we can set its flags */
+    _netif_thread = (thread_t *)thread_get(thread_getpid());
+
+#ifdef MODULE_GNRC_SIXLOWPAN
+    /* we disable fragmentation for this device, as the L2CAP layer takes care
+     * of this */
+    _nimble_netif->sixlo.max_frag_size = 0;
+#endif
+}
+
+static int _send_pkt(nimble_netif_conn_t *conn, gnrc_pktsnip_t *pkt)
+{
+    int res;
+    int num_bytes = 0;
+
+    if (conn->coc == NULL) {
+        return NIMBLE_NETIF_DEVERR;
+    }
+
+    /* copy the data into a newly allocated mbuf */
+    struct os_mbuf *sdu = os_mbuf_get_pkthdr(&_mbuf_pool, 0);
+    if (sdu == NULL) {
+        return NIMBLE_NETIF_NOMEM;
+    }
+    while (pkt) {
+        res = os_mbuf_append(sdu, pkt->data, pkt->size);
+        if (res != 0) {
+            os_mbuf_free_chain(sdu);
+            return NIMBLE_NETIF_NOMEM;
+        }
+        num_bytes += (int)pkt->size;
+        pkt = pkt->next;
+    }
+
+    /* send packet via the given L2CAP COC */
+    do {
+        res = ble_l2cap_send(conn->coc, sdu);
```

Comment:
It doesn't look like this function returns negative `errno`. Please map.

Reply:
it does. anyway, the return value is not passed on, so this is function internal in this particular place